l3build.1
.TH l3build 1 "2024-02-08"
.SH NAME
l3build \- Checking and building packages
.SH SYNOPSIS
l3build <target> [<options>] [<names>]
.SH DESCRIPTION
The l3build system is a Lua script for building TeX packages, with particular
emphasis on regression testing. It is written in cross-platform Lua code, so
can be used by any modern TeX distribution with the texlua interpreter. A
package for building with l3build can be written in any TeX dialect; its
defaults are set up for LaTeX packages written in the DocStrip style.
.PP
The standard l3build targets are:
.IP check
Runs all automated tests
.IP clean
Cleans out directory tree
.IP ctan
Creates CTAN-ready archive
.IP doc
Typesets all documentation files
.IP install
Installs files into the local texmf tree
.IP manifest
Creates a manifest file
.IP save
Saves test validation log
.IP tag
Updates release tags in files
.IP uninstall
Uninstalls files from the local texmf tree
.IP unpack
Unpacks the source files into the build tree
.IP upload
Sends archive to CTAN for public release
.SH OPTIONS
Various options apply
.IP --config|-c
Sets the config(s) used for running tests
.IP --date
Sets the date to insert into sources
.IP --debug
Runs target in debug mode
.IP --dirty
Skips cleaning up the test area
.IP --dry-run
Dry run for install or upload
.IP --email
Email address of CTAN uploader
.IP --engine|-e
Sets the engine(s) to use for running test
.IP --epoch
Sets the epoch for tests and typesetting
.IP --file|-F
Takes the upload announcement from the given file
.IP --first
Name of first test to run
.IP --full
Installs all files
.IP --halt-on-error|-H
Stops running tests after the first failure
.IP --help|-h
Prints help message and exits
.IP --last
Name of last test to run
.IP --message|-m
Text for upload announcement message
.IP --quiet|-q
Suppresses TeX output when unpacking
.IP --rerun
Skips setup: simply reruns tests
.IP --show-log-on-error
Shows the full log of the failure with 'halt-on-error'
.IP --show-saves|-S
Shows the invocation to update failing .tlg files
.IP --shuffle
Shuffles order of tests
.IP --stdengine|-s
Run tests with the std engine (config dependent)
.IP --texmfhome
Location of user texmf tree
.IP --version
Prints version information and exits
.SH BUGS
.SH AUTHOR
The LaTeX Project (latex-team@latex-project.org)
.PP
Please log issues on the GitHub homepage:
https://github.com/latex3/l3build/issues.
.SH SEE ALSO
See l3build.pdf for more details.


CHANGELOG.md
# Changelog
All notable changes to the `l3build` bundle since the start of 2018
will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
this project uses date-based 'snapshot' version identifiers.

## [Unreleased]

## [2024-02-08]

### Changed
- Extend normalization of line numbers to include those wrapped by LaTeX in
  text `on line ...`
- Suppress `l3msg` message wrapping

## [2024-01-18]

### Added
- Switch `-s|--stdengine` to run a set of tests only with the standard engine
  even where this varies between configs (issue \#343)

### Removed
- Switch `--force|-f`

## [2024-01-09]

### Fixed
- Type of return value of `runtest_tasks()` in doc
- Print failures correctly when these occur in multiple configurations
  including the core (`build`) one

## [2024-01-04]

### Changed
- Throw warnings on unknown doc name(s)
- Always execute `runtest_tasks()` if set (issue \#327)
- Print failures correctly when these occur in multiple configurations

## [2023-12-15]

### Fixed
- Test for blank `runtest_tasks()` (issue \#327)

## [2023-12-13-2]

- Correct setup for script running in TeX Live

## [2023-12-13]

### Fixed
- Syntax warning on Windows with some test setups

## [2023-12-12]

### Changed
- Document default value of `ctanpkg` as a valid lua expression
- Improve log for failed checks with no diff files
- Document full syntaxes of `\SHOWFILE` and `\ASSERT(STR)`

### Fixed
- Short-circuit `check --rerun` if `testdir` doesn't exist
- Retain errorlevel on Windows during `check` target

## [2023-11-01]

### Changed
- Drop duplicate backslashes in doc

### Fixed
- Support non-ASCII filenames that fall within the system codepage on Windows
  (see \#122)

## [2023-09-13]

### Added
- Document ConTeXt as supported `checkformat`

### Changed
- Extend version string normalisation during checks
  (see issue \#96)

- Extend excludefiles to cover `build.lua` (see \#286)
### Fixed
- Return passing errorlevel if BibTeX issues warnings
  (see \#260)
- Respect `excludefiles` when doing (local) installation

## [2023-09-07]

### Changed
- Refine `-utc` support
- Apply `checkopts` in addition to engine-specific options

## [2023-09-05]

### Changed
- Initialise the random seed with the current time so `--shuffle` produces different orders each run.
- Normalise more `luaotfload` cache lines
- Apply `-utc` switch for LuaTeX when using a fixed epoch value

## [2023-07-20]

- Set `-kanji-internal=euc` when building pLaTeX

## [2023-07-17]

### Changed
- Improve stdout "Running l3build with target ..."
- Quote configuration name used in stdout
- Update one leftover outdated doc for `unpackexe`: defaults to `pdftex`
- Building pLaTeX format now uses e-upTeX engine
- Normalize more `luaotfload` path data (see issue \#301)
- Update ConTeXt settings to allow for LuaTeX and LuaMetaTeX runs
- Improve doc for default `stdengine`

### Fixed
- Avoid setting `TEMXFCNF` for ConTeXt (issue \#232)

## [2023-03-27]

### Fixed
- All LuaTeX `.tlg` files were wrongly considered not engine-specific.
  Introduced in #292 which tried to fix #291.

## [2023-03-22]

### Changed
- Default value of `maxprintline` is now `9999`
  (may require `.tlg` updates: see docs)

### Fixed
- Apply needed luatex-specific log normalization, even when `--rerun` is used
  (issue \#291)

## [2023-03-08]

### Changed
- Generalize normalization of ghostscript version in PDF-based tests
- Include UNIX timestamps in generated ZIP files
- Normalize pdfTeX `.enc` file loading

### Fixed
- Ensure when used, value of `ps2pdfopts` is surrounded by a space on both sides

## [2023-02-26]

### Changed
- Run engine sanity check per config

### Fixed
- Restore epoch settings for `dvitopdf()`
- Use plural form of variable `ps2pdfopts` consistently in code and doc, and
  retain compatibility with singular form `ps2pdfopt` (issue \#275)
- Remove the last trace of dropped variable `stdconfig`

## [2023-02-20]

### Changed
- Unify `testdir` by dropping possibly trailing `.lua` passed to `--config`

### Fixed
- Ensure directories `testdir` and `resultdir` exist when `--dirty` is set
- epoch settings with xetex

## [2023-02-16]

### Changed
- Drop a redundant setup line for upTeX
- Normalise more Lua stack trace data (may require `.tlg` rebuild)

### Fixed
- Ensure `texmf.cnf` work correctly for `dvips`

## [2022-11-10]

### Changed
- Suppress (new) LaTeX version data at end of `.log`

### Fixed
- Allow for local override of `ctanupload` variable

## [2022-09-15]

### Fixed
- Copying of nested directories

## [2022-04-19]

### Changed
- Normalise GhostScript version in PDF-based tests
- Sort list of names of difference files for failing tests.

## [2022-04-12]

### Added
- Basic support for `make4ht`

### Changed
- Support `bidi` version string in `\special` lines (closes \#226)

## [2022-03-15]

### Changed
- When `\pdfmeta_set_regression_data:` is defined it is used
  to set metadata
- Support multiple configurations in bundles

### Fixed
- Correctly normalize luaotfload font cache path

## [2022-02-24]

### Fixed
- Creation of subdirectories in TDS structures on Unix-like systems

- use `form-string` rather than `form` for all curl fields to avoid
  misinterpreting leading `@` or `<` eg a description starting `<p>`

- Check the boolean value returned by executing shell commands in
  `l3build-upload` and throw an error if this is false. This fixes
  the issue that previously "validation successful" was reported
  if `curl` failed.

### Changed
- Documentation of how to validate an upload

## [2021-12-14]

### Fixed
- Use correct name for `options` table in multi configuration management code

## [2021-12-09]

### Added
- Support for pre-constructed TDS-style sources (variable `tdsdirs`)
- Support for injection of tokens using `specialformats`

### Changed
- If multiple configurations are present, let `l3build clean` run
  on all of them by default. (issue \#214)

## [2021-12-06]

### Fixed

l3build.lua
#!/usr/bin/env texlua

--[[

File l3build.lua Copyright (C) 2014-2024 The LaTeX Project

It may be distributed and/or modified under the conditions of the
LaTeX Project Public License (LPPL), either version 1.3c of this
license or (at your option) any later version.  The latest version
of this license is in the file

   https://www.latex-project.org/lppl.txt

This file is part of the "l3build bundle" (The Work in LPPL)
and all files in that bundle must be distributed together.

-----------------------------------------------------------------------

The development version of the bundle can be found at

   https://github.com/latex3/l3build

for those people who are interested.

--]]

-- Version information
release_date = "2023-12-15"

-- File operations are aided by the LuaFileSystem module
local lfs = require("lfs")

-- Local access to functions

local ipairs           = ipairs
local insert           = table.insert
local lookup           = kpse.lookup
local match            = string.match
local gsub             = string.gsub
local next             = next
local print            = print
local exit             = os.exit
local open             = io.open

-- l3build setup and functions
kpse.set_program_name("kpsewhich")
build_kpse_path = match(lookup("l3build.lua"),"(.*[/])")
local function build_require(s)
  require(lookup("l3build-"..s..".lua", { path = build_kpse_path } ) )
end

-- Minimal code to do basic checks
build_require("arguments")
build_require("help")

build_require("file-functions")
build_require("typesetting")
build_require("aux")
build_require("clean")
build_require("check")
build_require("ctan")
build_require("install")
build_require("unpack")
build_require("manifest")
build_require("manifest-setup")
build_require("tagging")
build_require("upload")
build_require("stdmain")

-- This has to come after stdmain(),
-- and that has to come after the functions are defined
if options["target"] == "help" then
  help()
  exit(0)
elseif options["target"] == "version" then
  version()
  exit(0)
end

-- Look for some configuration details
if fileexists("build.lua") then
  dofile("build.lua")
else
  print("Error: Cannot find configuration build.lua")
  exit(1)
end

-- Load standard settings for variables:
-- comes after any user versions
build_require("variables")

-- Ensure that directories are 'space safe'
maindir       = escapepath(maindir)
docfiledir    = escapepath(docfiledir)
sourcefiledir = escapepath(sourcefiledir)
supportdir    = escapepath(supportdir)
testfiledir   = escapepath(testfiledir)
testsuppdir   = escapepath(testsuppdir)
builddir      = escapepath(builddir)
distribdir    = escapepath(distribdir)
localdir      = escapepath(localdir)
resultdir     = escapepath(resultdir)
testdir       = escapepath(testdir)
typesetdir    = escapepath(typesetdir)
unpackdir     = escapepath(unpackdir)

-- Tidy up the epoch setting
-- Force an epoch if set at the command line
-- Must be done after loading variables, etc.
if options["epoch"] then
  epoch           = options["epoch"]
  forcecheckepoch = true
  forcedocepoch   = true
end
epoch = normalise_epoch(epoch)
-- LuaTeX needs the `-utc` option
if forcecheckepoch then
  if next(specialformats) and next(specialformats.latex)
    and next (specialformats.latex.luatex) then
    local options = specialformats.latex.luatex.options
    specialformats.latex.luatex.options = (options and (options .. " ") or "") .. "-utc"
  end
  if next(specialformats) and next(specialformats["latex-dev"])
    and next (specialformats["latex-dev"].luatex) then
    local options = specialformats["latex-dev"].luatex.options
    specialformats["latex-dev"].luatex.options = (options and (options .. " ") or "") .. "-utc"
  end
end
if forcedocepoch then
  if match(typesetexe,"luatex") or match(typesetexe,"lualatex") then
    typesetopts = typsetopts .. " -utc"
  end
end

--
-- Deal with multiple configs for tests
--

-- When we have specific files to deal with, only use explicit configs
-- (or just the default one)
if options["names"] then
  checkconfigs = options["config"] or {"build"}
else
  checkconfigs = options["config"] or checkconfigs
end

if #checkconfigs > 1 then
  if options["target"] == "check" or options["target"] == "bundlecheck" then
    local errorlevel = 0
    local failed = { }
    for i = 1, #checkconfigs do
      options["config"] = {checkconfigs[i]}
      errorlevel = call({"."}, "check", options)
      if errorlevel ~= 0 then
        if options["halt-on-error"] then
          exit(1)
        else
          insert(failed,checkconfigs[i])
        end
      end
    end
    if next(failed) then
      for _,config in ipairs(failed) do
        checkdiff(config)
      end
      if options["show-saves"] then
        local savecmds, recheckcmds = "", ""
        for _,config in ipairs(failed) do
          local testdir = testdir
          if config ~= "build" then
            testdir = testdir .. "-" .. config
          end
          local f = open(testdir .. "/.savecommands")
          if not f then
            print("Error: Cannot find save commands for configuration \"" ..
              config .. "\"")
            exit(2)
          end
          for line in f:lines() do
            if line == "" then break end
            savecmds = savecmds .. "  " .. line .. "\n"
          end
          for line in f:lines() do
            recheckcmds = recheckcmds .. "  " .. line .. "\n"
          end
          f:close()
        end
        print"To regenerate the test files, run\n"
        print(savecmds)
        if recheckcmds ~= "" and #checkengines ~= 1 then
          print"To detect engine-specific differences, run after that\n"
          print(recheckcmds)
        end
      end
      exit(1)
    else
      -- Avoid running the 'main' set of tests twice
      exit(0)
    end
  elseif options["target"] == "clean" then
    local failure
    for i = 1, #checkconfigs do
      options["config"] = {checkconfigs[i]}
      failure = 0 ~= call({"."}, "clean", options) or failure
    end
    exit(failure and 1 or 0)
  end
end
if #checkconfigs == 1 and
  (options["target"] == "check" or options["target"] == "save" or options["target"] == "clean") then
  if checkconfigs[1] == "build" then
    -- Sanity check for default config
    check_engines("build.lua")
  else
    local configname  = gsub(checkconfigs[1], "%.lua$", "")
    local config = "./" .. configname .. ".lua"
    if fileexists(config) then
      local savedtestfiledir = testfiledir
      dofile(config)
      -- Sanity check for non-default config
      check_engines(configname .. ".lua")
      testdir = testdir .. "-" .. configname
      -- Reset testsuppdir if required
      if savedtestfiledir ~= testfiledir and
        testsuppdir == savedtestfiledir .. "/support" then
        testsuppdir = testfiledir .. "/support"
      end
    else
      print("Error: Cannot find configuration \"" ..  configname .. ".lua\"")
      exit(1)
    end
  end
end

-- Call the main function
main(options["target"], options["names"])
